Doc/tools/check-warnings.py

```diff
@@ -24,28 +26,177 @@
+)
 
+# Regex pattern to match the line numbers in a Git unified diff
+DIFF_PATTERN = re.compile(
+    r"^@@ -(?P<linea>\d+)(?:,(?P<removed>\d+))? \+(?P<lineb>\d+)(?:,(?P<added>\d+))? @@",
+    flags=re.MULTILINE,
+)
 
-def check_and_annotate(warnings: list[str], files_to_check: str) -> None:
+
+def get_diff_files(ref_a: str, ref_b: str, filter_mode: str = "") -> set[Path]:
+    """List the files changed between two Git refs, filtered by change type."""
+    added_files_result = subprocess.run(
+        [
+            "git",
+            "diff",
+            f"--diff-filter={filter_mode}",
+            "--name-only",
+            f"{ref_a}...{ref_b}",
+            "--",
+        ],
+        stdout=subprocess.PIPE,
+        check=True,
+        text=True,
+        encoding="UTF-8",
+    )
+
+    added_files = added_files_result.stdout.strip().split("\n")
+    return {Path(file.strip()) for file in added_files if file.strip()}
+
+
+def get_diff_lines(ref_a: str, ref_b: str, file: Path) -> list[int]:
+    """List the lines changed between two Git refs for a specific file."""
+    diff_output = subprocess.run(
+        [
+            "git",
+            "diff",
+            "--unified=0",
+            f"{ref_a}...{ref_b}",
+            "--",
+            str(file),
+        ],
+        stdout=subprocess.PIPE,
+        check=True,
+        text=True,
+        encoding="UTF-8",
+    )
+
+    # Scrape line offsets + lengths from diff and convert to line numbers
+    line_matches = DIFF_PATTERN.finditer(diff_output.stdout)
+    line_match_values = [
+        line_match.groupdict(default=1) for line_match in line_matches
+    ]
+    line_ints = [
+        (int(match_value["lineb"]), int(match_value["added"]))
+        for match_value in line_match_values
+    ]
+    line_ranges = [
+        range(line_b, line_b + added) for line_b, added in line_ints
+    ]
+    line_numbers = list(itertools.chain(*line_ranges))
+
+    return line_numbers
+
+
+def get_para_line_numbers(file_obj: TextIO) -> list[list[int]]:
+    """Get the line numbers of text in a file object, grouped by paragraph."""
+    paragraphs = []
+    prev_line = None
+    for lineno, line in enumerate(file_obj):
+        lineno = lineno + 1
+        if prev_line is None or (line.strip() and not prev_line.strip()):
+            paragraph = [lineno - 1]
+            paragraphs.append(paragraph)
+        paragraph.append(lineno)
+        prev_line = line
+    return paragraphs
+
+
+def filter_and_parse_warnings(
+    warnings: list[str], files: set[Path]
+) -> list[re.Match[str]]:
+    """Get the warnings matching passed files and parse them with regex."""
+    filtered_warnings = [
+        warning
+        for warning in warnings
+        if any(str(file) in warning for file in files)
+    ]
+    warning_matches = [
+        WARNING_PATTERN.fullmatch(warning.strip())
+        for warning in filtered_warnings
+    ]
+    non_null_matches = [warning for warning in warning_matches if warning]
```

Comment:
```suggestion
    non_null_matches = list(filter(None, warning_matches))
```

Reply:
Is there a benefit here to using the functional constructs here, aside from being a bit shorter? The rest of this script heavily uses the more Pythonic comprehension syntax, so it seems a bit out of place to me to use the functional syntax in this one spot just to save a few characters...

Reply:
To (over-)indulge my curiosity, microbenchmarking it, seems like the `filter` approach does have a performance advantage, at least on 3.11 ([PEP 709][(https://peps.python.org/pep-0709/) in Python 3.12 is asserted to speed up comprehensions by 2x or more, particularly in cases like this where the ratio of overhead to actual work is maximized). Tested on the full list of nearly 6000 warnings, I got these results:

```ipython
%timeit non_null_matches = [warning for warning in warning_matches if warning]
576 µs ± 28.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit non_null_matches = list(filter(None, warning_matches))
163 µs ± 3.64 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

That's 3.5x faster. On the other hand, in absolute terms this is a sub-millisecond speedup (on my 10 year old laptop) on a line that is only run twice and represents an upper bound theoretical case as in reality it only actually runs on warnings from added and modified files, which is nearly always a small fraction of the total (on the order of tens to hundreds), and will continue to decrease over time. On a more realistic case adding and modifying a few files with a total of 40 warnings, the numbers look like:

```ipython
%timeit non_null_matches = [warning for warning in warning_matches if warning]
4.39 µs ± 210 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

%timeit non_null_matches = list(filter(None, warning_matches))
1.65 µs ± 23.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

The relative speedup is the same, but now we're only saving at most couple microseconds per run, which decreases over time due to Python optimizations, improved CI performance and decreased number of warnings. For comparison, that's around 20 million times less than the speedup with fetch/checkout from my changes above, to the point where at a generous estimate of 100 CI runs per day, every day, it would need to run for 550 years to save the same amount of time as a single run with the optimized checkout. Not to mention run for over 10 000 years in the time I spent indulging my curiosity here...but hope it was at least interesting and informative :joy:

Thank you for attending my TED talk, and I leave you with a profound quote:

"Something something something premature optimization is the root of all evil" — Winston Churchill